### ISSUE TITLE
An ApplyElementWise method that returns a column 

### ISSUE THREAD

**Issue by zHaytam** [dotnet/corefxlab#2805]
I'm trying out the `Microsoft.Data.Analysis` and was looking for a method that applies a `Func<,>` but returns the result as an array of values (maybe not efficient) or a column, just like in Pandas.

Here's what I'm currently doing:
```csharp
public static PrimitiveDataFrameColumn<TResult> Apply<T, TResult>(this PrimitiveDataFrameColumn<T> column, 
    Func<T, TResult> func) 
    where T : unmanaged
    where TResult : unmanaged
{
    var resultColumn = new PrimitiveDataFrameColumn<TResult>(string.Empty, 0);

    foreach (var row in column)
        resultColumn.Append(func(row.Value));

    return resultColumn;
}
```

Example usage:
```csharp
var birthdayColumn = df["Birthday"] as PrimitiveDataFrameColumn<DateTime>;
var currentYear = DateTime.Now.Year;
df["Age"] = birthdayColumn.Apply(d => currentYear - d.Year);
df.PrettyPrint();
```

Thank you!

**Comment by pgovind** [comment]
Hey @zHaytam, have you looked at https://github.com/dotnet/corefxlab/blob/master/src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.cs#L488? The `ApplyElementwise` API does something very similar to what you want. The only difference is that it works in place. To do what you want, you can do:

``` csharp
df["Age"] = df["Birthday"].Clone().ApplyElementwise((d, index) => return currentYear - d.Year));
```

**Comment by zHaytam** [comment]
Hello @pgovind, the problem with `ApplyElementWise` is that the output type is the same as the input. My birthday column is of type `DateTime` while the age column, I want it to be `int` :/

**Comment by pgovind** [comment]
Ah, my bad. I didn't see the return type. Yup, you're right. This is on my list of new APIs to add to `PrimitiveDataFrameColumn`. Would you be interested in putting up a PR perhaps? We could get this API in for the next preview :)

**Comment by zHaytam** [comment]
I would gladly try. I'm pretty sure the example I gave is far from being the best way to do it, I'm assuming it needs to be implemented directly in the Container?

**Comment by pgovind** [comment]
Yup. I would define the API in `PrimitiveDataFrameColumn`, maybe create the `returnColumn` in the API and then implement the loop inside `PrimitiveDataFrameColumnContainer`.

**Comment by zHaytam** [comment]
The same loop using the IEnumerable? I thought I would need to use the buffers in the container directly, which I understand for setting a value (like ApplyElementWise) but not for creating a new column out of it :/

**Comment by pgovind** [comment]
Right, that's why we'd define the `resultColumn` outside. Something like this ought to work:

``` csharp
// Define this in PrimitiveDataFrameColumn.cs

        public PrimitiveDataFrameColumn<TResult> Apply<TResult>(Func<T?, TResult?> func)
            where TResult : unmanaged
        {
            PrimitiveDataFrameColumn<TResult> resultColumn = new PrimitiveDataFrameColumn<TResult>("Result", Length);
            _columnContainer.Apply(func, resultColumn._columnContainer);
            return resultColumn;
        }

// Define this in PrimitiveColumnContainer.cs
        public void Apply<TResult>(Func<T?, TResult?> func, PrimitiveColumnContainer<TResult> resultContainer)
            where TResult : unmanaged
        {
            for (int b = 0; b < Buffers.Count; b++)
            {
                ReadOnlyDataFrameBuffer<T> buffer = Buffers[b];
                long prevLength = checked(Buffers[0].Length * b);
                DataFrameBuffer<T> mutableBuffer = DataFrameBuffer<T>.GetMutableBuffer(buffer);
                Buffers[b] = mutableBuffer;
                Span<T> span = mutableBuffer.Span;
                DataFrameBuffer<byte> mutableNullBitMapBuffer = DataFrameBuffer<byte>.GetMutableBuffer(NullBitMapBuffers[b]);
                NullBitMapBuffers[b] = mutableNullBitMapBuffer;
                Span<byte> nullBitMapSpan = mutableNullBitMapBuffer.Span;


                // Get Span<TResult> resultSpan = resultColumnMutableBuffer.Span similar to the above lines. We've assumed equals column lengths, so this should be straightforward
                // Similarly, get the resultColumnMutableNullBitMapBuffer
                for (int i = 0; i < span.Length; i++)
                {
                    long curIndex = i + prevLength;
                    bool isValid = IsValid(nullBitMapSpan, i);
                    TResult? value = func(isValid ? span[i] : default(T?));
                    resultSpan[i] = value.GetValueOrDefault();
                    SetValidityBit(resultColumnNullBitMapSpan, i, value != null);
                }
            }
        }


```

**Comment by zHaytam** [comment]
I have created the PR if you have the time to look at it.

**Comment by pgovind** [comment]
Closed with #2807
